core/trino-main/src/main/java/io/trino/execution/scheduler/StageTaskSourceFactory.java

```diff
@@ -418,102 +427,113 @@ public HashDistributionTaskSource(
-                }
-            }
+                        List<TaskDescriptor> result = postprocessTasks(partitionTasks.build());
+                        finished = true;
+                        return result;
+                    },
+                    directExecutor());
+        }
 
-            Map<Integer, Multimap<PlanNodeId, ExchangeSourceHandle>> partitionToExchangeSourceHandlesMap = new HashMap<>();
-            for (Map.Entry<PlanNodeId, ExchangeSourceHandle> entry : partitionedExchangeSourceHandles.entries()) {
-                PlanNodeId planNodeId = entry.getKey();
-                ExchangeSourceHandle handle = entry.getValue();
-                int partition = handle.getPartitionId();
-                Multimap<PlanNodeId, ExchangeSourceHandle> partitionSourceHandles = partitionToExchangeSourceHandlesMap.computeIfAbsent(partition, (p) -> ArrayListMultimap.create());
-                partitionSourceHandles.put(planNodeId, handle);
-            }
+        CompletableFuture<LoadedSplits> loadSplits(PlanNodeId planNodeId, SplitSource splitSource, List<Split> loadedSplits)
+        {
+            CompletableFuture<SplitBatch> splitBatchFuture = toCompletableFuture(splitSource.getNextBatch(NOT_PARTITIONED, Lifespan.taskWide(), splitBatchSize));
 
-            int taskPartitionId = 0;
-            ImmutableList.Builder<TaskDescriptor> partitionTasks = ImmutableList.builder();
-            for (Integer partition : union(partitionToSplitsMap.keySet(), partitionToExchangeSourceHandlesMap.keySet())) {
-                ListMultimap<PlanNodeId, Split> splits = partitionToSplitsMap.getOrDefault(partition, ImmutableListMultimap.of());
-                ListMultimap<PlanNodeId, ExchangeSourceHandle> exchangeSourceHandles = ImmutableListMultimap.<PlanNodeId, ExchangeSourceHandle>builder()
-                        .putAll(partitionToExchangeSourceHandlesMap.getOrDefault(partition, ImmutableMultimap.of()))
-                        // replicated exchange source will be added in postprocessTasks below
-                        .build();
-                Set<HostAddress> hostRequirement = partitionToNodeMap.get(partition);
-                partitionTasks.add(new TaskDescriptor(taskPartitionId++, splits, exchangeSourceHandles, new NodeRequirements(catalogRequirement, hostRequirement, taskMemory)));
-            }
+            long start = System.nanoTime();
+            return splitBatchFuture.thenCompose(splitBatch -> {
+                getSplitTimeRecorder.accept(start);
+                loadedSplits.addAll(splitBatch.getSplits());
 
-            List<TaskDescriptor> result = postprocessTasks(partitionTasks.build());
+                if (splitBatch.isLastBatch()) {
+                    splitSource.close();
+                    return completedFuture(new LoadedSplits(loadedSplits, planNodeId));
+                }
 
-            finished = true;
-            return result;
+                return loadSplits(planNodeId, splitSource, loadedSplits);
```

Comment:
Please refrain from recusing. Recursion may create a rather deep recursion stack (depends on the total number of splits and the number of splits returns as part of a single batch). e.g.:  
![image](https://user-images.githubusercontent.com/5570988/167726487-d581c2a6-6ec3-4513-bd81-4b51e4105a7f.png)


The stack of callbacks will be retained until the very last feature completes. This may result in retaining more objects / memory than necessary as well as complex object graph and hardly predictable performance when deep recursion stack is created. I would strongly suggest re implementing it in an iterative manner. Please take a look at `AbstractFuture`.

Reply:
I've made changes to avoid chain of completable futures now and implemented AbstractFuture. It still uses recursion though, not sure how to avoid that entirely.

Reply:
You would not have recursion if you did not use `directExecutore` but using multiple threads for what essentially is a loop is also suboptimal...

Reply:
I think I messed up with naming. Basically what i meant is future chaining. There's not really a way to avoid recursion. Though it would still be better to execute callbacks on a thread pool (not on a direct executor) to avoid deep recursion stacks.